colossalai/testing/utils.py

```diff
@@ -11,42 +12,121 @@ def parameterize(argument: str, values: List[Any]):
+        # > xavier: hello
+        # > xavier: bye
+        # > xavier: stop
+        # > davis: hello
+        # > davis: bye
+        # > davis: stop
+    
+    Args:
+        argument (str): the name of the argument to parameterize
+        values (List[Any]): a list of values to iterate for this argument
     """
 
     def _wrapper(func):
+
         def _execute_function_by_param(**kwargs):
             for val in values:
                 arg_map = {argument: val}
                 partial_func = partial(func, **arg_map)
                 partial_func(**kwargs)
+
         return _execute_function_by_param
+
+    return _wrapper
+
+
+def rerun_on_exception(exception_type: Exception = Exception, pattern: str = None, max_try: int = 5) -> Callable:
+    """
+    A decorator on a function to re-run when an exception occurs.
+
+    Usage::
+
+        # rerun for all kinds of exception
+        @rerun_on_exception()
+        def test_method():
+            print('hey')
+            raise RuntimeError('Address already in use')
+
+        # rerun for RuntimeError only
+        @rerun_on_exception(exception_type=RuntimeError)
+        def test_method():
+            print('hey')
+            raise RuntimeError('Address already in use')
+
+        # rerun for maximum 10 times if Runtime error occurs
+        @rerun_on_exception(exception_type=RuntimeError, max_try=10)
+        def test_method():
+            print('hey')
+            raise RuntimeError('Address already in use')
+        
+        # rerun for infinite times if Runtime error occurs
+        @rerun_on_exception(exception_type=RuntimeError, max_try=None)
+        def test_method():
+            print('hey')
+            raise RuntimeError('Address already in use')
+        
+        # rerun only the exception message is matched with pattern
+        # for infinite times if Runtime error occurs
+        @rerun_on_exception(exception_type=RuntimeError, pattern="^Address.*$")
+        def test_method():
+            print('hey')
+            raise RuntimeError('Address already in use')
+
+    Args:
+        exception_type (Exception, Optional): The type of exception to detect for rerun
+        pattern (str, Optional): The pattern to match the exception message. 
+            If the pattern is not None and matches the exception message,
+            the exception will be detected for rerun
+        max_try (int, Optional): Maximum reruns for this function. The default value is 5. 
+            If max_try is None, it will rerun foreven if exception keeps occurings
+    """
+
+    def _wrapper(func):
+
+        def _run_until_success(*args, **kwargs):
+            try_count = 0
+            assert max_try is None or isinstance(max_try, int), \
+                f'Expected max_try to be None or int, but got {type(max_try)}'
+
+            while max_try is None or try_count < max_try:
+                try:
+                    try_count += 1
+                    func(*args, **kwargs)
```

Comment:
What if the func has already handled the exception. Then no exception will be caught here.
Does the wrapper only work for test func without exception handling?

Reply:
If the exception is handled by internal code, then we should not catch it. We only need to handle those which make the program stop.